### PR TITLE
store content in keymap instead of hashtbl

### PR DESCRIPTION
**Pull request by vbmithr**
This preserves field order (not in the spec, but some programs rely on
this)

### CODE REVIEW

File: lib/pb.ml [hunk 1]
```diff
@@ -244,25 +246,23 @@ let read_unknown : Wire_type.t -> string Angstrom.t =
...
 
-let read_contents : (key, string list) Hashtbl.t Angstrom.t =
-  let add_one : (key, string list) Hashtbl.t -> key -> string -> unit =
+let read_contents : string list KeyMap.t Angstrom.t =
+  let add_one : string list KeyMap.t -> key -> string -> string list KeyMap.t =
     fun h k v ->
-      match Hashtbl.find h k with
-      | exception Not_found -> Hashtbl.add h k [v]
-      | vs -> Hashtbl.add h k (v :: vs)
+      match KeyMap.find_opt k h with
```
Comment (yallop) [inline lib/pb.ml]: Could you switch this (and the other occurrences of `find_opt`) to `KeyMap.find` to keep things working on OCaml 4.03 and 4.04?

**Comment by vbmithr** [comment]
On 31/07/2019 17:10, Jeremy Yallop wrote:
> Could you switch this (and the other occurrences of |find_opt|) to
> |KeyMap.find| to keep things working on OCaml 4.03 and 4.04?

Done.

**Comment by yallop** [comment]
Thank you!

**Comment by yallop** [comment]
Two more things to address, and then I think we can probably merge.

1. It would be better to consider all (i.e. both) fields in the comparison function in `KeyMap`: as it is, fields with the same key number but different wire types will be silently considered equal, which seems fragile.

2. Please use the `MoreLabels` version of `Map` to keep the existing convention with `iter`

**Comment by vbmithr** [comment]
Done.

**Pull request merged**